File: frontend/src/api/employeeDirectory.ts
import { apiFetch } from "./client";

export type VacationPeriod = {
  start: string;
  end: string;
};

export type WorkScheduleKind = "five_two" | "shift" | "two_two";
/** Пол сотрудника: для 5/2 из него считаются 8 ч или 7.2 ч */
export type EmployeeGender = "male" | "female" | "unspecified";

export type EmployeeDirectoryRowOut = {
  id: string;
  full_name: string;
  email: string;
  is_active: boolean;
  birth_date: string | null;
  position: { id: string; name: string; slug: string } | null;
  systems: { id: string; name: string; slug: string }[];
  exam_electrical_passed: boolean;
  exam_electrical_date: string | null;
  exam_electrical_valid_to: string | null;
  pass_has: boolean;
  pass_number: string | null;
  pass_valid_from: string | null;
  pass_valid_to: string | null;
  notes: string | null;
  vacation_periods: VacationPeriod[];
  work_schedule_kind: WorkScheduleKind;
  gender: EmployeeGender;
};

export type EmployeeDirectoryFilters = {
  search?: string;
  /** Хотя бы одна из систем (ИЛИ) */
  system_ids?: string[];
  /** Хотя бы одна из должностей (ИЛИ) */
  position_ids?: string[];
  exam_electrical_passed?: boolean;
  pass_has?: boolean;
  expiring_in_days?: number;
  expired_only?: boolean;
  include_inactive_users?: boolean;
  gender?: EmployeeGender;
  work_schedule_kind?: WorkScheduleKind;
};

export type EmployeeDirectoryPatch = {
  birth_date?: string | null;
  position_id?: string | null;
  system_ids?: string[];
  exam_electrical_passed?: boolean;
  exam_electrical_date?: string | null;
  exam_electrical_valid_to?: string | null;
  pass_has?: boolean;
  pass_number?: string | null;
  pass_valid_from?: string | null;
  pass_valid_to?: string | null;
  notes?: string | null;
  vacation_periods?: VacationPeriod[];
  work_schedule_kind?: WorkScheduleKind;
  gender?: EmployeeGender;
};

export async function listEmployeeDirectory(
  filters?: EmployeeDirectoryFilters,
): Promise<EmployeeDirectoryRowOut[]> {
  const sp = new URLSearchParams();
  if (filters?.search) sp.set("search", filters.search);
  if (filters?.system_ids?.length) {
    for (const id of filters.system_ids) sp.append("system_ids", id);
  }
  if (filters?.position_ids?.length) {
    for (const id of filters.position_ids) sp.append("position_ids", id);
  }
  if (filters?.exam_electrical_passed !== undefined) {
    sp.set("exam_electrical_passed", String(filters.exam_electrical_passed));
  }
  if (filters?.pass_has !== undefined) sp.set("pass_has", String(filters.pass_has));
  if (filters?.expiring_in_days !== undefined) sp.set("expiring_in_days", String(filters.expiring_in_days));
  if (filters?.expired_only) sp.set("expired_only", "true");
  if (filters?.include_inactive_users) sp.set("include_inactive_users", "true");
  if (filters?.gender) sp.set("gender", filters.gender);
  if (filters?.work_schedule_kind) sp.set("work_schedule_kind", filters.work_schedule_kind);
  const q = sp.toString();
  return apiFetch<EmployeeDirectoryRowOut[]>(`/api/v1/employee-directory${q ? `?${q}` : ""}`);
}

export async function getEmployeeDirectoryUser(userId: string): Promise<EmployeeDirectoryRowOut> {
  return apiFetch<EmployeeDirectoryRowOut>(`/api/v1/employee-directory/${userId}`);
}

export async function patchEmployeeDirectory(
  userId: string,
  body: EmployeeDirectoryPatch,
): Promise<EmployeeDirectoryRowOut> {
  return apiFetch<EmployeeDirectoryRowOut>(`/api/v1/employee-directory/${userId}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export type EmployeeDirectoryBulkProfilePatch = {
  work_schedule_kind?: WorkScheduleKind;
  gender?: EmployeeGender;
  position_id?: string | null;
  system_ids?: string[];
};

export async function bulkEmployeeDirectoryProfile(body: {
  user_ids: string[];
  patch: EmployeeDirectoryBulkProfilePatch;
}): Promise<{ updated: number }> {
  return apiFetch<{ updated: number }>("/api/v1/employee-directory/bulk-profile", {
    method: "POST",
    body: JSON.stringify(body),
  });
}

File: backend/app/schemas/employee_directory.py
import uuid
from datetime import date
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.schemas.position import PositionBrief
from app.schemas.system import SystemBrief


class VacationPeriodOut(BaseModel):
    start: date
    end: date
    kind: Literal["vacation", "study", "sick"] = "vacation"


class VacationPeriodIn(BaseModel):
    start: date
    end: date
    kind: Literal["vacation", "study", "sick"] = "vacation"

    @model_validator(mode="after")
    def check_range(self) -> "VacationPeriodIn":
        if self.end < self.start:
            raise ValueError("end must be >= start")
        return self


class EmployeeDirectoryRowOut(BaseModel):
    id: uuid.UUID
    full_name: str
    email: str
    is_active: bool
    birth_date: date | None = None
    position: PositionBrief | None = None
    systems: list[SystemBrief] = []

    exam_electrical_passed: bool = False
    exam_electrical_date: date | None = None
    exam_electrical_valid_to: date | None = None
    exam_electrical_group: str | None = None
    exam_electrical_certificate_number: str | None = None

    pass_has: bool = False
    pass_number: str | None = None
    pass_valid_from: date | None = None
    pass_valid_to: date | None = None

    notes: str | None = None
    vacation_periods: list[VacationPeriodOut] = Field(default_factory=list)
    work_schedule_kind: Literal["five_two", "shift", "two_two"] = "five_two"
    gender: Literal["male", "female", "unspecified"] = "unspecified"
    is_remote: bool = False
    work_address: str | None = None
    is_field_worker: bool = False
    position_assigned_at: date | None = None
    personnel_number: str | None = None


class EmployeeDirectoryPatch(BaseModel):
    birth_date: date | None = None
    position_id: uuid.UUID | None = None
    system_ids: list[uuid.UUID] | None = Field(None, description="Полная замена списка производственных систем.")

    exam_electrical_passed: bool | None = None
    exam_electrical_date: date | None = None
    exam_electrical_valid_to: date | None = None
    exam_electrical_group: str | None = Field(None, max_length=8)
    exam_electrical_certificate_number: str | None = Field(None, max_length=128)
    pass_has: bool | None = None
    pass_number: str | None = None
    pass_valid_from: date | None = None
    pass_valid_to: date | None = None
    notes: str | None = None
    vacation_periods: list[VacationPeriodIn] | None = Field(
        None,
        description="Полная замена списка периодов отпуска; до 24 интервалов. Учитывается при автозаполнении графика.",
    )
    work_schedule_kind: Literal["five_two", "shift", "two_two"] | None = None
    gender: Literal["male", "female", "unspecified"] | None = None
    is_remote: bool | None = None
    work_address: str | None = Field(None, max_length=512)
    is_field_worker: bool | None = None
    position_assigned_at: date | None = None
    personnel_number: str | None = Field(None, max_length=64)

    @model_validator(mode="after")
    def limit_vacation_periods(self) -> "EmployeeDirectoryPatch":
        if self.vacation_periods is not None and len(self.vacation_periods) > 24:
            raise ValueError("At most 24 vacation periods")
        return self


class EmployeeDirectoryBulkProfilePatch(BaseModel):
    """Кадровые поля для массового обновления (как столбцы справочника, кроме уникальных)."""

    work_schedule_kind: Literal["five_two", "shift", "two_two"] | None = None
    gender: Literal["male", "female", "unspecified"] | None = None
    position_id: uuid.UUID | None = None
    system_ids: list[uuid.UUID] | None = Field(None, description="Полная замена списка производственных систем.")
    is_remote: bool | None = None
    is_field_worker: bool | None = None
    work_address: str | None = Field(None, max_length=512)
    position_assigned_at: date | None = None


class EmployeeDirectoryBulkProfileIn(BaseModel):
    user_ids: list[uuid.UUID] = Field(..., min_length=1, max_length=2000)
    patch: EmployeeDirectoryBulkProfilePatch

    @model_validator(mode="after")
    def patch_not_empty(self) -> "EmployeeDirectoryBulkProfileIn":
        data = self.patch.model_dump(exclude_unset=True)
        if not data:
            raise ValueError("В patch нужно указать хотя бы одно поле")
        return self


class EmployeeDirectoryBulkProfileOut(BaseModel):
    updated: int
